R/buffer.R

```diff
@@ -306,3 +306,21 @@ as_raw <- function(buffer, client = pjrt_client(), row_major) {
   check_client(client)
   impl_client_buffer_to_raw(client, buffer, row_major = row_major)
 }
+
+#' Device of a PJRTBuffer
+#'
+#' @description
+#' Get the device of a [`PJRTBuffer`][pjrt_buffer].
+#'
+#' @template param_buffer
+#' @return `PJRTDevice`
+#' @export
+pjrt_device <- function(buffer) {
```

Comment:
By just reading the code it's getting hard to understand what are the main arguments for `pjrt_` functions. Sometimes it's a client, sometimes a program or a buffer.

Maybe we want `buffer$device` to work instead? Same for `buffer$element_type`? `device(buffer)` could be nice too (like we do `dim(buffer)`) but maybe to hard to discover.

I think we should reserve `pjrt_` for functions that create the main PJRT objects and make the methods that extract attributes with some different interface.

Not blocking this PR though, we can always change this later when we have consilidated our ideas.

Reply:
I like your suggestion of keeping the API "functional" with device(buffer) and element_type(buffer). I have implemented it.
I solved discoverability by adding a section "Extractors" that e.g. lists device() and element_type() for pjrt_buffer. We can still iterate on it afterwards. One name I am not sure about is pjrt_execute(), because it does not really create a pjrt object, but it's a central function, so I left it at pjrt_execute for now.